masp_primitives/src/transaction/components/amount.rs

```diff
@@ -143,180 +220,397 @@ impl Amount<AssetType> {
+        ValueSum(comps)
     }
 }
 
-impl<Unit: Hash + Ord + BorshSerialize + BorshDeserialize + Clone> Sub<&Amount<Unit>>
-    for Amount<Unit>
+impl<Unit, Magnitude> Sub<&ValueSum<Unit, Magnitude>> for ValueSum<Unit, Magnitude>
+where
+    Unit: Hash + Ord + BorshSerialize + BorshDeserialize + Clone,
+    Magnitude: BorshSerialize + BorshDeserialize + PartialEq + Eq + Copy + Default + CheckedSub,
 {
-    type Output = Self;
+    type Output = ValueSum<Unit, Magnitude>;
 
-    fn sub(mut self, rhs: &Self) -> Self {
-        self -= rhs;
-        self
+    fn sub(self, rhs: &ValueSum<Unit, Magnitude>) -> Self::Output {
+        let mut comps = self.0.clone();
+        for (atype, amount) in rhs.components() {
+            comps.insert(
+                atype.clone(),
+                self.get(atype)
+                    .checked_sub(amount)
+                    .expect("overflow detected"),
+            );
+        }
+        comps.retain(|_, v| *v != Magnitude::default());
+        ValueSum(comps)
     }
 }
 
-impl<Unit: Hash + Ord + BorshSerialize + BorshDeserialize + Clone> Sub<Amount<Unit>>
-    for Amount<Unit>
+impl<Unit, Magnitude> Sub<ValueSum<Unit, Magnitude>> for ValueSum<Unit, Magnitude>
+where
+    Unit: Hash + Ord + BorshSerialize + BorshDeserialize + Clone,
+    Magnitude: BorshSerialize + BorshDeserialize + PartialEq + Eq + Copy + Default + CheckedSub,
 {
-    type Output = Self;
+    type Output = ValueSum<Unit, Magnitude>;
 
-    fn sub(mut self, rhs: Self) -> Self {
-        self -= &rhs;
-        self
+    fn sub(self, rhs: ValueSum<Unit, Magnitude>) -> Self::Output {
+        self - &rhs
     }
 }
 
-impl<Unit: Hash + Ord + BorshSerialize + BorshDeserialize + Clone> Sum for Amount<Unit> {
+impl<Unit, Magnitude> Sum for ValueSum<Unit, Magnitude>
+where
+    Unit: Hash + Ord + BorshSerialize + BorshDeserialize + Clone,
+    Magnitude: BorshSerialize + BorshDeserialize + PartialEq + Eq + Copy + Default + PartialOrd,
+    Self: Add<Output = Self>,
+{
     fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
         iter.fold(Self::zero(), Add::add)
     }
 }
 
+/// Workaround for the blanket implementation of TryFrom
+pub struct TryFromNt<X>(pub X);
+
+impl<Unit, Magnitude, Output> TryFrom<TryFromNt<ValueSum<Unit, Magnitude>>>
+    for ValueSum<Unit, Output>
+where
+    Unit: Hash + Ord + BorshSerialize + BorshDeserialize + Clone,
+    Magnitude: BorshSerialize + BorshDeserialize + PartialEq + Eq + Copy,
+    Output: BorshSerialize + BorshDeserialize + PartialEq + Eq + Copy + TryFrom<Magnitude>,
+{
+    type Error = <Output as TryFrom<Magnitude>>::Error;
+
+    fn try_from(x: TryFromNt<ValueSum<Unit, Magnitude>>) -> Result<Self, Self::Error> {
+        let mut comps = BTreeMap::new();
+        for (atype, amount) in x.0 .0 {
+            comps.insert(atype, amount.try_into()?);
+        }
+        Ok(Self(comps))
+    }
+}
+
+/// Workaround for the blanket implementation of TryFrom
+pub struct FromNt<X>(pub X);
```

Comment:
Nit: This type isn't clear downstream. Can we give it a different name? I don't know the best, but something like `ConvertFrom`? The `Nt` confused me when I was reading the PR to namada.

Reply:
Understood. I've removed this wrapper type entirely and opted for `from_sum` and `try_from_sum` instance methods instead. Should be simpler this way.